File: rust/src/control_lamp.rs
extern crate rppal;

use rppal::gpio::{Gpio};

pub fn control_lamp(state: bool) -> Result<(), rppal::gpio::Error> {
    // Set up the GPIO pin connected to the relay module as an output pin
    let mut relay_pin = Gpio::new()?.get(17)?.into_output();

    if state {
        // Set the GPIO pin to high (on)
        relay_pin.set_high();
        println!("Lamp turned on");
    } else {
        // Set the GPIO pin to low (off)
        relay_pin.set_low();
        println!("Lamp turned off");
    }

    Ok(())
}

File: heating-controller/control_lamp.py
import RPi.GPIO as GPIO

def control_lamp(state: str):
    """Controls a lamp.

    Args:
        state: The state to set the lamp to. Valid values are 'on' and 'off'.
    """
    # Set up the GPIO pin for the relay
    relay_pin = 17
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(relay_pin, GPIO.OUT)
    
    

    if state == 'on':
        # Turn the relay on
        GPIO.output(relay_pin, GPIO.HIGH)
    elif state == 'off':
        # Turn the relay off
        GPIO.output(relay_pin, GPIO.LOW)
    else:
        raise ValueError(f'Invalid state: {state}')